### PR TITLE
Stop recursion on trivial replacement

### LINKED ISSUE #97817: Subgraph_rewriter tracks replacement nodes incorrectly for trivial replacement pattern
### 🐛 Describe the bug

Pattern replacement behaves incorrectly when the replacement pattern maps inputs to outputs - such a pattern can be used to replace redundant code. However, current code in `torch.fx.subgraph_rewriter._replace_pattern` causes the list of replacement nodes to include the entire graph before that node, resulting in an exponential slowdown due to recursive calls traversing the entire graph multiple times. This slowdown is trivial in the example code but very significant in production models.

To reproduce:
```python
import torch
from torch.fx.subgraph_rewriter import replace_pattern_with_filters
from torch.fx.experimental.proxy_tensor import make_fx

def model(x):
    x = (x+2) - 3 + x.T
    return x.unsqueeze(0).squeeze()

def redundant_pattern(input):
    x = torch.ops.aten.unsqueeze.default(input, 0)
    return torch.ops.aten.squeeze.default(x)
def replacement(input):
    return input

x = torch.randn(2,2)
gm = make_fx(model)(x)
gm.graph.print_tabular()
matches = replace_pattern_with_filters(gm, redundant_pattern, replacement, [])
print(matches[0].replacements)  # Expected only [add_2], instead see all parents of add_2
```

My proposed fix is to add a check in `_replace_pattern` to prevent the call to `get_replacement_nodes`:
```python
        for ret_node in copied_returning_nodes:
            if ret_node in match.placeholder_nodes:
                replacement_nodes.append(ret_node)
            else:
                get_replacement_nodes(ret_node)
```

### Versions

```
PyTorch version: 2.0.0a0+gitdf69368
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.27

Python version: 3.9.15 (main, Nov 24 2022, 14:31:59)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.15.0-96-generic-x86_64-with-glibc2.27
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              112
On-line CPU(s) list: 0-111
Thread(s) per core:  2
Core(s) per socket:  28
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz
Stepping:            4
CPU MHz:             2627.712
CPU max MHz:         3800.0000
CPU min MHz:         1000.0000
BogoMIPS:            5000.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            39424K
NUMA node0 CPU(s):   0-27,56-83
NUMA node1 CPU(s):   28-55,84-111
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke md_clear flush_l1d

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.2.1
[pip3] flake8==6.0.0
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] torch==2.0.0a0+git085e81a
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.0a0+7f778fc
[pip3] torchmetrics==0.11.3
[pip3] torchrec-nightly==2023.3.8
[pip3] torchtext==0.14.0a0+e2b27f9
[pip3] torchvision==0.14.0a0+5ce4506
[pip3] vector-quantize-pytorch==1.1.1
[conda] blas                      1.0                         mkl  
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.2.1                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-include               2022.1.0           h06a4308_224  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.2                   pypi_0    pypi
[conda] numpy-base                1.23.5           py39h31eccc5_0  
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] torch                     2.0.0a0+git085e81a           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.0a0+7f778fc          pypi_0    pypi
[conda] torchmetrics              0.11.3                   pypi_0    pypi
[conda] torchrec-nightly          2023.3.8                 pypi_0    pypi
[conda] torchtext                 0.14.0a0+e2b27f9          pypi_0    pypi
[conda] torchvision               0.14.0a0+5ce4506          pypi_0    pypi
[conda] vector-quantize-pytorch   1.1.1                    pypi_0    pypi
```

cc @ezyang @SherlockNoMad @soumith @EikanWang @jgong5 @wenzhe-nrv @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by frost-intel**
Pattern replacement behaves incorrectly when the replacement pattern maps inputs to outputs (such a pattern can be used to replace redundant code). However, current code in `torch.fx.subgraph_rewriter._replace_pattern` causes the list of replacement nodes to include the entire graph before that node, resulting in an exponential slowdown due to recursive calls traversing the entire graph multiple times.

The proposed fix is to add a check in `_replace_pattern` to prevent the call to `get_replacement_nodes`:
```python
        for ret_node in copied_returning_nodes:
            if ret_node in match.placeholder_nodes:
                replacement_nodes.append(ret_node)
            else:
                get_replacement_nodes(ret_node)
```

Fixes #97817 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/fx/subgraph_rewriter.py [hunk 1]
```diff
@@ -309,7 +309,10 @@ def get_replacement_nodes(curr_node: Node):
             replacement_nodes.append(curr_node)
 
         for ret_node in copied_returning_nodes:
-            get_replacement_nodes(ret_node)
+            if ret_node in match.placeholder_nodes:
+                replacement_nodes.append(ret_node)
```
Comment (jgong5) [inline torch/fx/subgraph_rewriter.py:313]: Not sure if I understand correctly, but is it possible that we still need to scan the args of the `ret_node` for `replacement_nodes` even if it is the placeholder of a match?

Comment (frost-intel) [reply torch/fx/subgraph_rewriter.py:313]: `match.placeholder_nodes` are nodes in the target graph which correspond to placeholders (inputs) in the search pattern. This means that if `ret_node` is in `match.placeholder_nodes`, none of it's args are part of the search pattern, so none of them will be replaced.

Comment (angelayi) [inline torch/fx/subgraph_rewriter.py:313]: I think this should be an empty list, since there are no nodes that are created in the replaced graph.

**Comment by frost-intel** [comment]
@angelayi Made requested changes, inputs will not be added to replacement nodes.

**Review by angelayi — APPROVED** [review]
Thanks!

**Pull request closed without merging**